libdnf/base/goal.cpp

```diff
@@ -1690,6 +1693,97 @@ void Goal::Impl::add_paths_to_goal() {
+        rpm::PackageQuery weak_query(base, rpm::PackageQuery::ExcludeFlags::APPLY_EXCLUDES);
+        libdnf::ResolveSpecSettings settings{
+            .with_nevra = true, .with_provides = false, .with_filenames = true, .with_binaries = false};
+        weak_query.resolve_pkg_spec(exclude_weak, settings, false);
+        weak_query.filter_available();
+        rpm_goal.add_exclude_from_weak(*weak_query.p_impl);
+    }
+}
+
+void Goal::Impl::autodetect_unsatisfied_installed_weak_dependencies() {
+    rpm::PackageQuery installed_query(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES);
+    installed_query.filter_installed();
+    if (installed_query.empty()) {
+        return;
+    }
+    rpm::PackageQuery base_query(base, rpm::PackageQuery::ExcludeFlags::APPLY_EXCLUDES);
+    rpm::ReldepList reldep_list(base);
+
+    std::vector<std::string> installed_names;
+    installed_names.reserve(installed_query.size());
+
+    // Investigate uninstalled recommends of installed packages
+    for (auto pkg : installed_query) {
+        installed_names.push_back(pkg.get_name());
+        for (auto recommend : pkg.get_recommends()) {
+            if (libdnf::rpm::Reldep::is_rich_dependency(recommend.to_string())) {
+                // Rich dependencies are skipped because they are too complicated to provide correct result
+                continue;
+            };
+            rpm::PackageQuery query(base_query);
+
+            //  There can be installed provider in different version or upgraded packed can recommend a different
+            //  version therefore ignore version and search only by reldep name
+            if (auto version = recommend.get_version(); version && strlen(version) > 0) {
+                auto & pool = get_rpm_pool(base);
+                Id id = pool.str2id(recommend.get_name(), 0);
+                reldep_list.add(rpm::ReldepId(id));
+                query.filter_provides(reldep_list);
+            } else {
+                reldep_list.add(recommend);
+                query.filter_provides(reldep_list);
+            };
+            reldep_list.clear();
+            // No providers of recommend => continue
+            if (query.empty()) {
+                continue;
+            }
+            rpm::PackageQuery test_installed(query);
+            test_installed.filter_installed();
+            // when there is not installed any provider of recommend, exclude it
+            if (test_installed.empty()) {
+                rpm_goal.add_exclude_from_weak(*query.p_impl);
+            }
+        }
+    }
+
+    // Investigate supplements of only available packages with a different name to installed packages
+    // We can use base_query, because it is not useful anymore
+    base_query.filter_name(installed_names, sack::QueryCmp::NEQ);
+    // We have to remove all installed packages from testing set
+    base_query -= installed_query;
+    rpm::PackageSet exclude_supplements(base);
+    for (auto pkg : base_query) {
```

Comment:
Better to use `const auto & pkg`?

Reply:
I was not sure who owns the package object. But I guess that reference should work, because the iterator should be owned by query.